modules/phase_field/src/utils/MisorientationAngleCalculator.C

```diff
@@ -0,0 +1,212 @@
+      q_num = {
+        { 1.000E+00,	 0.000E+00, 	 0.000E+00, 	 0.000E+00},
+        { 5.000E-01,	 5.000E-01, 	 5.000E-01, 	 5.000E-01},
+        {-5.000E-01,	 5.000E-01, 	 5.000E-01, 	 5.000E-01},
+        { 6.123E-17,	 7.071E-01, 	 7.071E-01, 	 8.660E-17},
+        {-7.071E-01,	 1.543E-16, 	 7.071E-01, 	 9.881E-17},
+        {-7.071E-01,	-7.071E-01, 	 2.343E-16, 	 1.221E-17},
+        { 7.071E-01,	 0.000E+00, 	 0.000E+00, 	 7.071E-01},
+        { 1.110E-16,	 7.071E-01, 	 3.925E-17, 	 7.071E-01},
+        {-7.071E-01,	 7.071E-01, 	 5.551E-17, 	 2.680E-16},
+        {-1.793E-17,	 1.000E+00, 	 8.865E-17, 	 1.045E-16},
+        {-5.000E-01,	 5.000E-01, 	 5.000E-01, 	-5.000E-01},
+        {-5.000E-01,	-5.000E-01, 	 5.000E-01, 	-5.000E-01},
+        { 6.123E-17,	 0.000E+00, 	 0.000E+00, 	 1.000E+00},
+        {-5.000E-01,	 5.000E-01, 	-5.000E-01, 	 5.000E-01},
+        {-5.000E-01,	 5.000E-01, 	-5.000E-01, 	-5.000E-01},
+        {-8.660E-17,	 7.071E-01, 	-7.071E-01, 	 6.123E-17},
+        {-1.421E-16,	 7.071E-01, 	-1.110E-16, 	-7.071E-01},
+        {-5.551E-17,	 1.910E-16, 	 7.071E-01, 	-7.071E-01},
+        {-7.071E-01,	 0.000E+00, 	 0.000E+00, 	 7.071E-01},
+        {-7.071E-01,	 7.177E-17, 	-7.071E-01, 	 1.340E-16},
+        {-3.005E-16,	 5.551E-17, 	-7.071E-01, 	-7.071E-01},
+        {-1.045E-16,	 1.009E-16, 	-1.000E+00, 	-1.793E-17},
+        { 5.000E-01,	 5.000E-01, 	-5.000E-01, 	-5.000E-01},
+        { 5.000E-01,	 5.000E-01, 	 5.000E-01, 	-5.000E-01}
+      }; // 24 symmetric for fcc
+
+    std::vector<QuatReal> q(q_num.size());
+    for (const auto i : index_range(q_num))
+    {
+      q[i].w() = q_num[i][0];
+      q[i].x() = q_num[i][1];
+      q[i].y() = q_num[i][2];
+      q[i].z() = q_num[i][3];
+    }
+    return q;
+  }
+
+  Real dotQuaternion(const QuatReal & o1, const QuatReal & o2, 
+                                              const std::vector<QuatReal> & qcs, 
+                                              const std::vector<QuatReal> & qss)
+  {
+    Real d = 0.0; // used to get misorientation angle
+    QuatReal mori = itimesQuaternion(o1, o2);
+
+    if (qss.size() <= 1)
+      d = dotOuterQuaternion(mori, qcs); 
+    else
+      d = mtimes2Quaternion(o1, qcs, o2); // mtimesQuaternion(qss,mtimesQuaternion(o1,qcs,0),1)
+
+    return d;
+  }
+
+  QuatReal itimesQuaternion(const QuatReal & q1, const QuatReal & q2)
+  {
+    Real a1, b1, c1, d1;
+    Real a2, b2, c2, d2;
+
+    a1 = q1.w(); b1 = -q1.x(); c1 = -q1.y(); d1 = -q1.z();
+    a2 = q2.w(); b2 = q2.x(); c2 = q2.y(); d2 = q2.z();
+
+    // standard algorithm
+    QuatReal q;
+    q.w() = a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2;
+    q.x() = b1 * a2 + a1 * b2 - d1 * c2 + c1 * d2;
+    q.y() = c1 * a2 + d1 * b2 + a1 * c2 - b1 * d2;
+    q.z() = d1 * a2 - c1 * b2 + b1 * c2 + a1 * d2;
+
+    return q;
+  }
+
+  Real dotOuterQuaternion(const QuatReal & rot1, const std::vector<QuatReal> & rot2)
+  {
+    std::vector<Real> d_vec(rot2.size());
```

Comment:
creating a local vector is very slow, you'd be better off to compute the largest element on the fly.

Reply:
Okay, you mean to use this calculation method?

```c++
    Real d = 0;
    Real temp = 0;
    for (const auto i : index_range(rot2))
    {
      temp = std::abs(rot1.w()*rot2[i].w() + rot1.x()*rot2[i].x() + rot1.y()*rot2[i].y() + rot1.z()*rot2[i].z()); // rot1 * rot2'
      if (temp > d)
        d = temp;
    }

    return d;
```